--- gpio-pcf857x-new.c	2026-03-04 10:31:06.026443079 +0800
+++ gpio-diodes.c	2026-03-04 10:47:39.534199741 +0800
@@ -5,6 +5,7 @@
  * Copyright (C) 2007 David Brownell
  */
 
+#include <linux/bitops.h>
 #include <linux/gpio/driver.h>
 #include <linux/i2c.h>
 #include <linux/interrupt.h>
@@ -31,6 +32,7 @@
 	{ "pca9675", 16 },
 	{ "max7328", 8 },
 	{ "max7329", 8 },
+	{ "pi4ioe5v96248", 48 },
 	{ }
 };
 MODULE_DEVICE_TABLE(i2c, pcf857x_id);
@@ -49,6 +51,7 @@
 	{ .compatible = "nxp,pca9675", (void *)16 },
 	{ .compatible = "maxim,max7328", (void *)8 },
 	{ .compatible = "maxim,max7329", (void *)8 },
+	{ .compatible = "diodes,pi4ioe5v96248", (void *)48 },
 	{ }
 };
 MODULE_DEVICE_TABLE(of, pcf857x_of_table);
@@ -69,31 +72,38 @@
 	struct gpio_chip	chip;
 	struct i2c_client	*client;
 	struct mutex		lock;		/* protect 'out' */
-	unsigned int		out;		/* software latch */
-	unsigned int		status;		/* current status */
-	unsigned int		irq_enabled;	/* enabled irqs */
+	u64			out;		/* software latch */
+	u64			status;		/* current status */
+	u64			irq_enabled;	/* enabled irqs */
 
-	int (*write)(struct i2c_client *client, unsigned int data);
-	int (*read)(struct i2c_client *client);
+	int (*write)(struct i2c_client *client, u64 data);
+	int (*read)(struct i2c_client *client, u64 *data);
 };
 
 /*-------------------------------------------------------------------------*/
 
 /* Talk to 8-bit I/O expander */
 
-static int i2c_write_le8(struct i2c_client *client, unsigned int data)
+static int i2c_write_le8(struct i2c_client *client, u64 data)
 {
 	return i2c_smbus_write_byte(client, data);
 }
 
-static int i2c_read_le8(struct i2c_client *client)
+static int i2c_read_le8(struct i2c_client *client, u64 *data)
 {
-	return i2c_smbus_read_byte(client);
+	int status;
+
+	status = i2c_smbus_read_byte(client);
+	if (status < 0)
+		return status;
+
+	*data = status;
+	return 0;
 }
 
 /* Talk to 16-bit I/O expander */
 
-static int i2c_write_le16(struct i2c_client *client, unsigned int word)
+static int i2c_write_le16(struct i2c_client *client, u64 word)
 {
 	u8 buf[2] = { word & 0xff, word >> 8, };
 	int status;
@@ -102,7 +112,7 @@
 	return (status < 0) ? status : 0;
 }
 
-static int i2c_read_le16(struct i2c_client *client)
+static int i2c_read_le16(struct i2c_client *client, u64 *data)
 {
 	u8 buf[2];
 	int status;
@@ -110,7 +120,41 @@
 	status = i2c_master_recv(client, buf, 2);
 	if (status < 0)
 		return status;
-	return (buf[1] << 8) | buf[0];
+
+	*data = (buf[1] << 8) | buf[0];
+	return 0;
+}
+
+/* Talk to 48-bit I/O expander */
+
+static int i2c_write_le48(struct i2c_client *client, u64 word)
+{
+	u8 buf[6] = {
+		word & 0xff,
+		(word >> 8) & 0xff,
+		(word >> 16) & 0xff,
+		(word >> 24) & 0xff,
+		(word >> 32) & 0xff,
+		(word >> 40) & 0xff,
+	};
+	int status;
+
+	status = i2c_master_send(client, buf, 6);
+	return (status < 0) ? status : 0;
+}
+
+static int i2c_read_le48(struct i2c_client *client, u64 *data)
+{
+	u8 buf[6];
+	int status;
+
+	status = i2c_master_recv(client, buf, 6);
+	if (status < 0)
+		return status;
+
+	*data = ((u64)buf[5] << 40) | ((u64)buf[4] << 32) | ((u64)buf[3] << 24) |
+		((u64)buf[2] << 16) | ((u64)buf[1] << 8) | (u64)buf[0];
+	return 0;
 }
 
 /*-------------------------------------------------------------------------*/
@@ -121,7 +165,7 @@
 	int status;
 
 	mutex_lock(&gpio->lock);
-	gpio->out |= (1 << offset);
+	gpio->out |= (1ULL << offset);
 	status = gpio->write(gpio->client, gpio->out);
 	mutex_unlock(&gpio->lock);
 
@@ -131,23 +175,26 @@
 static int pcf857x_get(struct gpio_chip *chip, unsigned int offset)
 {
 	struct pcf857x *gpio = gpiochip_get_data(chip);
-	int value;
+	u64 value;
+	int status;
 
-	value = gpio->read(gpio->client);
-	return (value < 0) ? value : !!(value & (1 << offset));
+	status = gpio->read(gpio->client, &value);
+	return (status < 0) ? status : !!(value & (1ULL << offset));
 }
 
 static int pcf857x_get_multiple(struct gpio_chip *chip, unsigned long *mask,
 				unsigned long *bits)
 {
 	struct pcf857x *gpio = gpiochip_get_data(chip);
-	int value = gpio->read(gpio->client);
+	u64 value;
+	int status;
 
-	if (value < 0)
-		return value;
+	status = gpio->read(gpio->client, &value);
+	if (status < 0)
+		return status;
 
 	*bits &= ~*mask;
-	*bits |= value & *mask;
+	*bits |= (unsigned long)value & *mask;
 
 	return 0;
 }
@@ -155,7 +202,7 @@
 static int pcf857x_output(struct gpio_chip *chip, unsigned int offset, int value)
 {
 	struct pcf857x *gpio = gpiochip_get_data(chip);
-	unsigned int bit = 1 << offset;
+	u64 bit = 1ULL << offset;
 	int status;
 
 	mutex_lock(&gpio->lock);
@@ -180,8 +227,8 @@
 	struct pcf857x *gpio = gpiochip_get_data(chip);
 
 	mutex_lock(&gpio->lock);
-	gpio->out &= ~*mask;
-	gpio->out |= *bits & *mask;
+	gpio->out &= ~((u64)*mask);
+	gpio->out |= (u64)(*bits & *mask);
 	gpio->write(gpio->client, gpio->out);
 	mutex_unlock(&gpio->lock);
 }
@@ -191,17 +238,21 @@
 static irqreturn_t pcf857x_irq(int irq, void *data)
 {
 	struct pcf857x *gpio = data;
-	unsigned long change, i, status;
+	u64 value;
+	unsigned long change, i;
+	int status;
 
-	status = gpio->read(gpio->client);
+	status = gpio->read(gpio->client, &value);
+	if (status < 0)
+		return IRQ_NONE;
 
 	/*
 	 * call the interrupt handler iff gpio is used as
 	 * interrupt source, just to avoid bad irqs
 	 */
 	mutex_lock(&gpio->lock);
-	change = (gpio->status ^ status) & gpio->irq_enabled;
-	gpio->status = status;
+	change = (unsigned long)((gpio->status ^ value) & gpio->irq_enabled);
+	gpio->status = value;
 	mutex_unlock(&gpio->lock);
 
 	for_each_set_bit(i, &change, gpio->chip.ngpio)
@@ -228,7 +279,7 @@
 	irq_hw_number_t hwirq = irqd_to_hwirq(data);
 
 	gpiochip_enable_irq(&gpio->chip, hwirq);
-	gpio->irq_enabled |= (1 << hwirq);
+	gpio->irq_enabled |= (1ULL << hwirq);
 }
 
 static void pcf857x_irq_disable(struct irq_data *data)
@@ -236,7 +287,7 @@
 	struct pcf857x *gpio = irq_data_get_irq_chip_data(data);
 	irq_hw_number_t hwirq = irqd_to_hwirq(data);
 
-	gpio->irq_enabled &= ~(1 << hwirq);
+	gpio->irq_enabled &= ~(1ULL << hwirq);
 	gpiochip_disable_irq(&gpio->chip, hwirq);
 }
 
@@ -273,11 +324,11 @@
 static int pcf857x_probe(struct i2c_client *client)
 {
 	struct pcf857x *gpio;
-	unsigned int n_latch = 0;
+	u64 n_latch = 0;
+	u64 value;
+	u32 t_latch = 0;
 	int status;
 
-	device_property_read_u32(&client->dev, "lines-initial-states", &n_latch);
-
 	/* Allocate, initialize, and register this gpio_chip. */
 	gpio = devm_kzalloc(&client->dev, sizeof(*gpio), GFP_KERNEL);
 	if (!gpio)
@@ -297,6 +348,15 @@
 	gpio->chip.direction_output	= pcf857x_output;
 	gpio->chip.ngpio		= (uintptr_t)i2c_get_match_data(client);
 
+	if (gpio->chip.ngpio == 48) {
+		device_property_read_u64(&client->dev, "lines-initial-states",
+					 &n_latch);
+	} else {
+		device_property_read_u32(&client->dev, "lines-initial-states",
+					 &t_latch);
+		n_latch = t_latch;
+	}
+
 	/* NOTE:  the OnSemi jlc1562b is also largely compatible with
 	 * these parts, notably for output.  It has a low-resolution
 	 * DAC instead of pin change IRQs; and its inputs can be the
@@ -318,7 +378,7 @@
 
 		/* fail if there's no chip present */
 		else
-			status = i2c_smbus_read_byte(client);
+			status = i2c_read_le8(client, &value);
 
 	/* '75/'75c addresses are 0x20..0x27, just like the '74;
 	 * the '75c doesn't have a current source pulling high.
@@ -335,8 +395,17 @@
 
 		/* fail if there's no chip present */
 		else
-			status = i2c_read_le16(client);
+			status = i2c_read_le16(client, &value);
+	} else if (gpio->chip.ngpio == 48) {
+		gpio->write	= i2c_write_le48;
+		gpio->read	= i2c_read_le48;
+
+		if (!i2c_check_functionality(client->adapter, I2C_FUNC_I2C))
+			status = -EIO;
 
+		/* fail if there's no chip present */
+		else
+			status = i2c_read_le48(client, &value);
 	} else {
 		dev_dbg(&client->dev, "unsupported number of gpios\n");
 		status = -EINVAL;
@@ -345,7 +414,9 @@
 	if (status < 0)
 		goto fail;
 
-	gpio->chip.label = client->name;
+	if (device_property_read_string(&client->dev, "label",
+					&gpio->chip.label))
+		gpio->chip.label = client->name;
 
 	gpio->client = client;
 	i2c_set_clientdata(client, gpio);
@@ -366,7 +437,10 @@
 	 * reset state.  Otherwise it flags pins to be driven low.
 	 */
 	gpio->out = ~n_latch;
-	gpio->status = gpio->read(gpio->client);
+	status = gpio->read(gpio->client, &value);
+	if (status < 0)
+		goto fail;
+	gpio->status = value;
 
 	/* Enable irqchip if we have an interrupt */
 	if (client->irq) {
@@ -410,7 +484,7 @@
 	struct pcf857x *gpio = i2c_get_clientdata(client);
 
 	/* Drive all the I/O lines high */
-	gpio->write(gpio->client, BIT(gpio->chip.ngpio) - 1);
+	gpio->write(gpio->client, BIT_ULL(gpio->chip.ngpio) - 1);
 }
 
 static struct i2c_driver pcf857x_driver = {
